src/java/org/apache/cassandra/index/sai/disk/vector/CompressedVectorRowIdIterator.java

```diff
@@ -0,0 +1,145 @@
+            this.appoximateScore = appoximateScore;
+        }
+
+        @Override
+        public int compareTo(ApproximateScore o)
+        {
+            // We always want scores in descending order
+            return Float.compare(o.appoximateScore, appoximateScore);
+        }
+    }
+
+    private final float[] queryVector;
+    // We use two PriorityQueues because we do not need an eager ordering of these results. Depending on how many
+    // sstables the query hits and the relative scores of vectors from those sstables, we may not need to return
+    // more than the first handful of scores.
+    // Priority queue with compressed vector scores
+    private final PriorityQueue<ApproximateScore> cvScoreQueue;
+    // Priority queue with full resolution scores
+    private final PriorityQueue<ScoredRowId> scoredRowIdQueue;
+    private final IntFunction<float[]> vectorForOrdinal;
+    private final VectorSimilarityFunction similarityFunction;
+    private final int topK;
+    private final int limit;
+    private final float threshold;
+
+    /**
+     * @param queryVector The query vector
+     * @param cvScoreQueue A priority queue of rows and their ordinal ordered by their approximate similarity scores
+     * @param vectorForOrdinal A function that returns the full resolution vector for a given ordinal
+     * @param similarityFunction The similarity function to compare full resolution vectors
+     * @param limit The query limit
+     * @param topK The number of vectors to resolve and score before returning results
+     * @param threshold The minimum similarity score to return
+     */
+    public CompressedVectorRowIdIterator(float[] queryVector,
+                                         PriorityQueue<ApproximateScore> cvScoreQueue,
+                                         IntFunction<float[]> vectorForOrdinal,
+                                         VectorSimilarityFunction similarityFunction,
+                                         int limit,
+                                         int topK,
+                                         float threshold)
+    {
+        this.queryVector = queryVector;
+        this.cvScoreQueue = cvScoreQueue;
+        this.scoredRowIdQueue = new PriorityQueue<>(topK, (a, b) -> Float.compare(b.score, a.score));
+        this.vectorForOrdinal = vectorForOrdinal;
+        this.similarityFunction = similarityFunction;
+        assert topK >= limit : "topK must be greater than or equal to limit. Found: " + topK + " < " + limit;
+        this.limit = limit;
+        this.topK = topK;
+        this.threshold = threshold;
+    }
+
+
+    @Override
+    public boolean hasNext()
+    {
+        // The priority queue is only valid for the first limit results. After that, we need to
+        // get the next `limit` results from the compressed vector graph, compute the full resolution scores,
+        // and add to the priority queue.
+        if (!scoredRowIdQueue.isEmpty() && !needsRefill())
```

Comment:
I don't follow the logic here -- I thought we can read all the exact-scored rows out of the exact-score PQ, before we need to add the next batch

Reply:
The logic is to refill the topK before we get too low so that we don't get odd behavior at the `k` and `k + 1` boundary. I thought this was the design we discussed on slack.

Reply:
Could be the virus talking -- I will revisit later.

Reply:
FWIW, the logic on this line was unnecessarily complex. Pushing up a fix.

Reply:
... I thought that what I said in Slack was, "you can't prevent odd behavior in the general case so let's not bother trying." :)

Reply:
I guess it wasn't in Slack, it was elsewhere on this PR.
> You can find a better score if you search deeper after 100 results, 1 result, or 0 results, there's no way to prevent this 100% without scanning the entire graph.

Reply:
... so not exactly this situation but the principle is similar -- there are no bounds on the on the actual similarity that you can predict from the approximate similarities, you can't return a smoothly decreasing sequence unless you exact-sort the whole thing up front which is what we're trying to avoid.

Reply:
... and I think this is not user-visible in any case?  Since we're still going to put the results back in PK order to send to the coordinator, and it's going to do a re-sort-by-score on its end.  (Not ideal of course, but maybe this is a silver lining?)

Reply:
> > so to resume you should remove the values returned in (3) from the list of candidates, and re-run steps 2 and 3. step 1 does not need to be re-run
>
> bonus points if you don't have to call getVector again for rows that were in the top k the first time but weren't returned (so k - limit rows)

That was the slack message I used for the design.

I am fine with going with a design where we exhaust the `exactScoreQueue` and then refill with the next topK essentially ignoring the `limit` value. This is not user-visible, as you point out and can easily be tuned later given more information about usage.

In the interest of time, I'll push up a commit simplifying the design.

Reply:
Ah, right, I remember this.  I'd say that my pattern-matching didn't fire because that was about accuracy, not preventing discontinuities but really my memory was just bad, sorry.

As penance I will clean this up to use Reranker.